esp-hal/src/lcd_cam/lcd/mod.rs

```diff
@@ -4,13 +4,566 @@
+    #[inline(always)]
+    pub fn set_blank_cycles(&self, fk_cycles: u32, bk_cycles: u32) {
+        self.lcd_cam.lcd_misc().modify(|_, w| {
+            w.lcd_bk_en()
+                .bit(fk_cycles > 0 || bk_cycles > 0)
+                .lcd_vfk_cyclelen()
+                .variant(fk_cycles.wrapping_sub(1) as _)
+                .lcd_vbk_cyclelen()
+                .variant(bk_cycles.wrapping_sub(1) as _)
+        });
+    }
+
+    /// Configure the data line width
+    /// This is essentially the amount of bits per pixel and thus either 8 or
+    /// 16.
+    #[inline(always)]
+    pub fn set_data_width(&self, width: BusWidth) {
+        self.lcd_cam
+            .lcd_user()
+            .modify(|_, w| w.lcd_2byte_en().bit(width == BusWidth::Bit16));
+    }
+
+    /// Configures whether the LCD peripheral continues the data phase as long
+    /// as the DMA has content to send.
+    ///
+    /// # Arguments
+    /// * `en` - `true`: The number of data cycles will be controlled by DMA
+    /// buffer size, instead of lcd_dout_cyclelen        `false`: The number
+    /// of data cycles will be controlled by lcd_dout_cyclelen
+    #[inline(always)]
+    pub fn enable_output_always_on(&self, enabled: bool) {
+        self.lcd_cam
+            .lcd_user()
+            .modify(|_, w| w.lcd_always_out_en().bit(enabled));
+    }
+
+    /// Start the LCD transaction
+    /// This will cause this LCD peripheral to start pushing data from the FIFO.
+    ///
+    /// # Warning
+    /// Make sure data in the FIFO is ready *before* calling this method,
+    /// otherwise garbage data will be pushed into the display.
+    #[inline(always)]
+    pub fn start(&self) {
+        // update parameters before actually starting the transaction
+        self.lcd_cam
+            .lcd_user()
+            .modify(|_, w| w.lcd_update().set_bit());
+        // start transaction
+        self.lcd_cam
+            .lcd_user()
+            .modify(|_, w| w.lcd_start().set_bit());
+    }
+
+    /// Stop the LCD transaction
+    #[inline(always)]
+    pub fn stop(&self) {
+        self.lcd_cam
+            .lcd_user()
+            .modify(|_, w| w.lcd_start().clear_bit());
+        self.lcd_cam
+            .lcd_user()
+            .modify(|_, w| w.lcd_update().set_bit());
+    }
+
+    /// Reset LCD TX controller and RGB/YUV converter
+    ///
+    /// # Warning
+    /// This does not reset the data currently contained in the TX FIFO buffer.
+    #[inline(always)]
+    pub fn reset(&self) {
+        self.lcd_cam
+            .lcd_user()
+            .modify(|_, w| w.lcd_reset().set_bit());
+    }
+
+    /// Reset the LCD peripheral's asynchronous TX FIFO buffer.
+    #[inline(always)]
+    pub fn fifo_reset(&self) {
+        self.lcd_cam
+            .lcd_misc()
+            .modify(|_, w| w.lcd_afifo_reset().set_bit());
+    }
+
+    /// (I8080-Mode) Set the level of the DC line for different transaction phases.
```

Comment:
If this is I8080 specific, why bother having this method here? This can stay in the I8080 driver.

Reply:
The "(I8080-Mode)" comment is just a "typically used by" which was added by me.
The way I understood it from a short conversation with an Espressif engineer and from the datasheet, there is not actually a `mode-specific` option. Even if you enable the chip's RGB-mode, some options (which are typically used for I8080 e.g.) would still work to cover certain use-cases.
That's why I followed their lcd_ll scheme: Because I'm not a big fan of making something as basic and as low in the system as a HAL too restrictive by encoding use-cases into the structure. IMHO, those interfaces never really age well.
If you move I8080-specific functions to the I8080 class, you effectively prevent anyone who really knows what they are doing from working properly with the hardware by doing expert things.

But in the end, I guess it's for @jessebraham and @bjoernQ to decide.

Reply:
I don't know if you've seen https://github.com/esp-rs/esp-hal/issues/1282 or https://github.com/esp-rs/esp-hal/discussions/31, but the general goal with the HAL as I see it is to provide restrictive set of APIs that enforce correct usage. It sounds like what you really want is a richer PAC (which I would kinda like too).

I haven't seen it discussed on this repo yet but I think there's plenty room for improvement when it comes to exotic use cases for a peripheral. The LCD_CAM was designed for LCDs and Cameras but really one could also use it for SPI, PARLIO, etc. The I8080 driver as is can even be used for MOTO6080 (though I couldn't find a good name to capture both so I went with the more popular one imo). And I think this is the point you're making about the lcd_ll scheme to provide flexibility without starting from scratch.

Whilst I 100% agree that a HAL should enable this flexibility, without a real use case to design for, it's impossible to design something sensible or review any design/refactor. Flexibility isn't free, so one needs to make sure they know what they're paying for and if it's worth it.

The current real use case I see here and I'm reviewing for is, "sharing code between the I8080 driver and the RGB driver".
With this in mind, I'm able to determine whether the method names make sense, whether functions like this one we're looking at (`set_dc_level`) that sets multiple registers at once should be split up or not, or whether the function is needed at all.
At the moment this `set_dc_level` has fantastic name and granularity for I8080 but for RGB it doesn't. For RGB it should have a different name and should only really care about `data_phase` (and *maybe* `idle_phase`, but in practice once the driver is started in RGB mode it typically needs to continuously run forever).
The only reason I'm able to make this judgement is because we have a real use case to discuss. If we want to handle every possible use case ever, then a richer PAC is simply the answer.

Just my opinion at least. I'll wait for the esp-hal maintainers to comment before I review further as I don't want to waste time and brain power making comments that the hal doesn't care for.